### ISSUE TITLE
Compiled .exe connects to unknown IP addresses

### ISSUE THREAD

**Issue by embeh** [pyinstaller/pyinstaller#8323]
Hi,

I am using pyinstaller 5.0.1 with Winpython 3.10.4.0 to compile my Python code into .exe on Windows 10 .

Due to a false-positive malware alarm, I looked into this more closely and found that the .exe contacts some unknown IP addresses and was wondering whether this is part of pyinstaller and/or some Microsoft tools used for it?

The IP addresses are 

192.229.211.108:80 (TCP)
20.99.185.48:443 (TCP)
23.34.172.81:443 (TCP)

In case it matters, these are the modules the code imports:

socket
time
serial
json
tkinter
tkinter.messagebox 
idlelib.tooltip 
argparse
os
winsound
math 
stupidArtnet

Any ideas where this is coming from?

Thank you.

**Comment by rokm** [comment]
What methodology did you use to determine these connection attempts and that they are coming from the executable (as opposed to, say, Windows smart screen or Windows defender when checking the executable)?

**Comment by rokm** [comment]
See for example: https://en.delphipraxis.net/topic/10070-delphi-ce-application-accesses-unknown-ips (since delphi program is very much unlikely to be PyInstaller-related).

**Comment by embeh** [comment]
> What methodology did you use to determine these connection attempts and that they are coming from the executable (as opposed to, say, Windows smart screen or Windows defender when checking the executable)?

I uploaded it to virustotal and they run the executable in multiple VMs/sandboxes.
So it seems to be quite similar to the delphi forum topic.

I have not yet tried to make a 'hello world' executable to see whether it does the same thing.

**Comment by bwoodsend** [comment]
> I have not yet tried to make a 'hello world' executable to see whether it does the same thing.

To be honest, you're just wasting our time then.

**Comment by embeh** [comment]
> > I have not yet tried to make a 'hello world' executable to see whether it does the same thing.
> 
> To be honest, you're just wasting our time then.

That is honest.

I will do a quick 'hello world" and report back.

**Comment by embeh** [comment]
Python code:
`print("hello world")`

Compiled into .exe using this command:
`pyinstaller helloworld.py --onefile --noconsole`

Result:
https://www.virustotal.com/gui/file/e701b39cdf2066d135645510f42d602f1ac166eaae9aa6869e1f88a22a58be29/behavior

This seems to contact the following IP addresses:
192.229.211.108:80 (TCP)
20.99.133.109:443 (TCP)
20.99.185.48:443 (TCP)
20.99.186.246:443 (TCP)
23.213.175.51:443 (TCP)

Thoughts?

**Comment by rokm** [comment]
Search for those IPs on google and see if they come up in similar context?

**Comment by embeh** [comment]
They do, for example here https://www.reddit.com/r/dotnet/comments/1213wc9/strange_network_communications/?rdt=50314

But that is referring to MS dotNet telemetry - is this something pyinstaller uses?

**Comment by rokm** [comment]
> But that is referring to MS dotNet telemetry - is this something pyinstaller uses?

Nope, because there's no .NET involved - the bootloader is written in C and the pre-compiled bootloader is built with Visual Studio C/C++ compiler.

**Comment by embeh** [comment]
Any chance the VS C++ compiler part is doing this then?

**Comment by rokm** [comment]
I'm not completely discounting that possibility - they did insert telemetry into generated executables at some point back with VS2015. But I think if they did it again, it would again make news. 

If you have Visual Studio available, you can try compiling a simple "hello world" C program, and upload it for analysis.

Here's my example:

```c
// main.c
#include <stdio.h>

int main(int argc, char **argv)
{
    printf("Hello world for VS-made executable!\n");
    return 0;
}
```

```
cl main.c /o:myprogram.exe
```

(FWIW, it's Visual Studio Community 2022 17.9.0).

https://www.virustotal.com/gui/file/7058467521e653e8f79f3972177b3ecb7d15c45ee64480057c0d4c44d97dcad9/behavior

So it's similar in this case as well:
192.229.211.108:80 (TCP)
20.99.184.37:443 (TCP)

**Comment by embeh** [comment]
Thanks for testing this. This looks like VS is doing some extra 'networking' then...
Not sure I like that but this is not a pyinstaller problem (unless there is an easy way to configure VS to not do that?)

**Comment by rokm** [comment]
I'm not even sure it's really VS / the executable. For all we know, it could be that first time a machine sees an executable, Windows smart screen, Windows defender, or whatever AV is there contacts its mothership with hash (or maybe even submit a sample - the defender has an option to disable automatic sample submission buried somewhere in the settings). Could be that this also happens in those sandboxes.

**Comment by rokm** [comment]
Either way, if you want to be sure, you should check the PyInstaller's bootloader code and rebuild the bootloader yourself, instead of using the prebuild variant (it might also help you with false positives, as then you won't use the exactly same bootloader executable as the majority of the world).

**Comment by embeh** [comment]
Just to wrap this up:

I just compiled the hello world code above with mingw-w64 into an .exe and uploaded that to virustotal.
https://www.virustotal.com/gui/file/7584fffaa201550ce760361d627dc0c084eaa1512addb920e824d5c95dc1b7a6/behavior

Result: 

IP addresses contacted:
192.229.211.108:80 (TCP)
20.99.133.109:443 (TCP)
20.99.185.48:443 (TCP)
20.99.186.246:443 (TCP)
23.219.78.176:443 (TCP)
a83f:8110:cce1:d301:e361:49b8:fa65:da01:53 (UDP)

**Comment by rokm** [comment]
Heh, and I thought our bootloaders have problem with false positives... I guess nothing is safe in a world where "hello world" programs are dangerous.

**Comment by embeh** [comment]
Looking at the dropped file from the MinGW example, this might be related to Windows Error Reporting...

**Comment by SynergyUSAllc** [comment]
Same problem here, Latest pyinstaller ver 6.4.0, one line python project, print("hello world"), generates 10 false positives:
( I only have the default Windows Defender and using VS2022 community )


https://www.virustotal.com/gui/file/26ca5018f3f61b586f666650d87ceb8b04e2222f8679ce81288c2265ebef25aa/detection

Sorry to sound pessimist, but this makes unfortunately the whole pyinstaller project obsolete.
And I am sure there are 100s of man hours worked so far
It's a shame to waste all those people effort.

For me I cannot post software on our web site that people will download and then Windows Defender pops up that its a trojan!

 I already uploaded my solution to microsoft for evaluation/analysis and waiting feedback from them.

Is there a solution to this issue ? 
Has someone found a solution to this ?

BTW when I digital signed my .exe app, the number of False Positive went down to 6:
https://www.virustotal.com/gui/file/7793eef6f20646b02c983cc457463e522a7c477e33439dcfc0280ae5913fedbd?nocache=1

I dont know if that helps someone.

How can we help the rest of the people facing the same issue ?
Is there a way to assist ?

**Comment by bwoodsend** [comment]
I think the IP addresses still being contacted with hello world + gcc is pretty conclusive that it's not the application itself that's to blame for these communications. On the flip side, simplewall isn't detecting any connections when I try run a PyInstaller application locally.

So all in all, I don't see anything we can act on in here.

**Comment by SynergyUSAllc** [comment]
I was able to solve the problem, and now my windows app is clean and VirusTotal reports zero threats:
https://www.virustotal.com/gui/file/f78e178ee49ae00579f2bbb9a5e12b74fe4715e7437ee81f538263b4d50f51e1/behavior

I hope it will also help other people, this is what I did:

I rebuilt the bootloader, but not using the VS2022 C compiler.
I used the MinGW-w64

Here are the bootloader nstructions on how to build in windows:
https://pyinstaller.org/en/v3.6/bootloader-building.html#building-for-windows

And this is where i used the compiler from:
https://winlibs.com/#download-release
(unzip to a root folder, and added bin folder to PATH)

As stated in the pyinstaller docs, I used 'python ./waf --gcc all' to built the bootloader.
I also uploaded the bootloader run.exe to V.T. and it was found clean.
(When I first tried to compile the bootloader using the VS2022 C compiler, V.T. reported that the bootloader run.exe was infected, so don't try that and save some time.)

My app looks clean now, 
my only problem is that I cannot digitalluy signed it, 
I get an error that the app is not a windows app (Where with earlier versions of Bootloader, I never had any issues adding my Digital Code Signature)

There are still two IP addresses in the V.T. sandboxes analysis, but the number of threats is zero.
